Realm/RLMResults.mm

```diff
@@ -563,6 +570,137 @@ - (RLMNotificationToken *)addNotificationBlock:(void (^)(RLMResults *, RLMCollec
+// The `timeout` param in subscribeWithName is a double and can't be nullable.
+// So there can't be a simple check for a null `timeout` in a single subscribeWithName method.
+// resolveSubscribeWithSyncMode holds the repeated switch statements. It returns a bool because some
+// cases fall through to a method calls which use a timeout.
+- (bool)checkEarlyReturnSubscribeWithSyncMode:(RLMWaitForSyncMode)waitForSyncMode
+                                name:(NSString *)name
+                             onQueue:(dispatch_queue_t _Nullable)queue
+                          completion:(RLMResultsCompletionBlock)completion {
+    RLMSyncSubscriptionSet *subscriptions = self.realm.subscriptions;
+    switch(waitForSyncMode) {
+        case RLMWaitForSyncModeOnCreation:
+            // If an existing named subscription matches the provided name and local query, return.
+            if (name) {
+                RLMSyncSubscription *sub = [subscriptions subscriptionWithName:name query:_results.get_query()];
+                if (sub != nil) {
+                    [self completeOnQueue:queue completion:completion error:nil];
+                    return true;
+                }
+            } else {
+                // otherwise check if an unnamed subscription already exists. Return if it does exist.
+                RLMSyncSubscription *sub = [subscriptions subscriptionWithQuery:_results.get_query()];
+                if (sub != nil && sub.name == nil) {
+                    [self completeOnQueue:queue completion:completion error:nil];
+                    return true;
+                }
+            }
+            // If no name was provided and no existing unnamed subscription matches.
+            // break and create new subscription later.
+            break;
+        case RLMWaitForSyncModeAlways:
+            // never returns early
+            break;
+        case RLMWaitForSyncModeNever:
+            // commit subscription synchronously and return.
+            [subscriptions update:^{
+                self.associatedSubscriptionId = [subscriptions addSubscriptionWithClassName:self.objectClassName
+                                                                           subscriptionName:name
+                                                                                      query:_results.get_query()
+                                                                             updateExisting:true];
+            }];
+            [self completeOnQueue:queue completion:completion error:nil];
+            return true;
+        }
+    return false;
+    }
+
+- (void)subscribeWithName:(NSString *_Nullable)name
+          waitForSyncMode:(RLMWaitForSyncMode)waitForSyncMode
+                  onQueue:(dispatch_queue_t _Nullable)queue
+               completion:(RLMResultsCompletionBlock)completion {
+    if ([self checkEarlyReturnSubscribeWithSyncMode:waitForSyncMode name:name onQueue:queue completion:completion]) {
+        return;
+    } else {
```

Comment:
The whole point of an early return is to not have the else block, and doing both is just sorta redundant extra code. `} else {` is also not the style used in our obj-c code.